### ISSUE TITLE
feature: load best model checkpoint

### ISSUE THREAD

**Issue by csarron** [huggingface/accelerate#418]
Hi, I was wondering if one can easily load the best saved model checkpoint (possibly with deepspeed) for evaluation/prediction after the training is finished. Similar to [what transformers Trainer did](https://github.com/huggingface/transformers/blob/ba286fe7d51db12ad663effac83bed8199dd7141/src/transformers/trainer.py#L1745):

```python
    def _load_best_model(self):
        logger.info(f"Loading best model from {self.state.best_model_checkpoint} (score: {self.state.best_metric}).")
        best_model_path = os.path.join(self.state.best_model_checkpoint, WEIGHTS_NAME)
        strict_load = is_sagemaker_mp_enabled()
        model = self.model_wrapped if is_sagemaker_mp_enabled() else self.model
        if os.path.exists(best_model_path):
            if self.deepspeed:
                # temp hack until Deepspeed fixes the problem with resume from an existing engine that did some stepping
                deepspeed_engine, optimizer, lr_scheduler = deepspeed_reinit(self)
                self.model = deepspeed_engine.module
                self.model_wrapped = deepspeed_engine
                self.deepspeed = deepspeed_engine
                self.optimizer = optimizer
                self.lr_scheduler = lr_scheduler
                self.deepspeed.load_checkpoint(
                    self.state.best_model_checkpoint, load_optimizer_states=True, load_lr_scheduler_states=True
                )
            else:
                # We load the model state dict on the CPU to avoid an OOM error.
                state_dict = torch.load(best_model_path, map_location="cpu")
                # If the model is on the GPU, it still works!
                load_result = model.load_state_dict(state_dict, strict=strict_load)

```

Do we have to reinit the deepspeed model in the accelerator library?

**Comment by sgugger** [comment]
We could had this in one of the feature example, but it's not something that will be directly in Accelerate.

cc @muellerzr if you're interested.

**Comment by pacman100** [comment]
Hello @csarron , yes, we can easily load the best saved model checkpoint (with deepspeed) for evaluation/prediction after the training is finished. And, no, we need not have to reinit the deepspeed model as per my experiments with DeepSpeed version (0.6.5). One important thing to note that the main purpose of `model.load_checkpoint()`  is to be able to **resume training from that instant again and not the case if one just wants to do evaluation/inference as it loads optimizer/scheduler states too**. Nonetheless, In PR #405 , we have put an example `examples/by_feature/deepspeed_with_config_support.py`. In that example, it is tracking the best model and loads it post training for evaluation. Below is the relevant code snippet doing that:

```python
# New Code #
def load_training_checkpoint(model, load_dir, tag=None, **kwargs):
    """Utility function for checkpointing model + optimizer dictionaries
    The main purpose for this is to be able to resume training from that instant again
    """
    _, checkpoint_state_dict = model.load_checkpoint(load_dir, tag=tag, **kwargs)
    epoch = checkpoint_state_dict["epoch"]
    last_global_step = checkpoint_state_dict["last_global_step"]
    del checkpoint_state_dict
    return (epoch, last_global_step)


# New Code #
# Loads the best checkpoint after the training is finished
if args.load_best_model:
    _, last_global_step = load_training_checkpoint(
        model,
        "/".join(best_metric_checkpoint.split("/")[:-1]),
        tag=best_metric_checkpoint.split("/")[-1],
        **{"load_optimizer_states": True, "load_lr_scheduler_states": True},
    )

    # New Code #
    # Evaluates using the best checkpoint
    perplexity, eval_loss = evaluate(args, model, eval_dataloader, accelerator, eval_dataset)
    logger.info(f"Best model metrics: perplexity: {perplexity} eval_loss: {eval_loss}")
    if perplexity != best_metric:
        raise AssertionError(
            f"Best metric {best_metric} does not match the metric {perplexity} of the loaded best model."
        )
```
For using model post training just to do inference/evaluation, please refer the `Saving and loading` section of the `deepspeed.mdx` doc file in the above PR (#405 ).
